File: routes/facts.py
from fastapi import APIRouter, Depends, HTTPException
from mongoengine import DoesNotExist, ValidationError
from models.models import Fact, Location, User, Tag
from utils.auth_utils import get_current_user
from typing import List
import logging
from schemas.facts import FactResponse, FactCreate

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/v3")

@router.get("/locations/{location_id}/facts", response_model=List[FactResponse])
def get_facts_by_location(location_id: str, current_user: User = Depends(get_current_user)):
    logger.info(f"Fetching facts for location ID: {location_id} by user ID: {current_user.id}")
    facts = Fact.objects(location_id=location_id)
    if not facts:
        logger.warning(f"No facts found for location ID: {location_id}")
        raise HTTPException(status_code=404, detail="No facts found for this location.")
    
    location = Location.objects.get(id=location_id)
    logger.info(f"Found {len(facts)} facts for location ID: {location_id}")
    return [FactResponse(
        id=str(fact.id),
        description=fact.description,  # Get description from Location model
        location_id=str(fact.location_id.id),
        user_id=str(current_user.id),
        created_at=fact.created_at
    ) for fact in facts]

@router.post("/locations/{location_id}/facts", response_model=FactResponse)
def create_fact(location_id: str, fact: FactCreate, current_user: User = Depends(get_current_user)):
    logger.info(f"Creating fact for location ID: {location_id} by user ID: {current_user.id}")
    try:
        location = Location.objects.get(id=location_id)
    except DoesNotExist:
        logger.warning(f"Location ID: {location_id} does not exist")
        raise HTTPException(status_code=404, detail="Location not found")

    new_fact = Fact(
        description=location.description,  # Get description from Location model
        location_id=location,
        user_id=current_user
    )
    new_fact.save()
    logger.info(f"Fact created with ID: {new_fact.id} for location ID: {location_id}")
    return FactResponse(
        id=str(new_fact.id),
        description=new_fact.description,
        location_id=str(new_fact.location_id.id),
        user_id=str(current_user.id),
        created_at=new_fact.created_at
    )

@router.delete("/facts/{fact_id}", status_code=204)
def delete_fact(fact_id: str, current_user: User = Depends(get_current_user)):
    logger.info(f"Attempting to delete fact with ID: {fact_id} by user ID: {current_user.id}")
    try:
        fact = Fact.objects.get(id=fact_id, user_id=current_user)
        fact.delete()
        logger.info(f"Fact with ID: {fact_id} deleted")
    except DoesNotExist:
        logger.warning(f"Fact ID: {fact_id} does not exist or not owned by user ID: {current_user.id}")
        raise HTTPException(status_code=404, detail="Fact not found")

@router.put("/locations/{location_id}/facts/{fact_id}", response_model=FactResponse)
def update_fact(location_id: str, fact_id: str, fact: FactCreate, current_user: User = Depends(get_current_user)):
    logger.info(f"Updating fact with ID: {fact_id} for location ID: {location_id} with new description: {fact.description} by user ID: {current_user.id}")
    try:
        location_exists = Location.objects.get(id=location_id)
    except DoesNotExist:
        logger.warning(f"Location ID {location_id} does not exist.")
        raise HTTPException(status_code=404, detail="Location not found.")

    try:
        fact_to_update = Fact.objects.get(id=fact_id, location_id=location_id)
        fact_to_update.update(description=fact.description)
        fact_to_update.reload()
        logger.info(f"Fact with ID {fact_id} updated successfully for location ID: {location_id}")
        return FactResponse(
            id=str(fact_to_update.id),
            description=fact_to_update.description,
            location_id=str(fact_to_update.location_id),
            user_id=str(current_user.id),
            created_at=fact_to_update.created_at
        )
    except DoesNotExist:
        logger.warning(f"Fact with ID {fact_id} not found for location ID {location_id}")
        raise HTTPException(status_code=404, detail="Fact not found.")
    except ValidationError as e:
        logger.error(f"Validation error updating fact: {str(e)}")
        raise HTTPException(status_code=400, detail="Invalid data provided.")
    except Exception as e:
        logger.error(f"Error updating fact: {str(e)}")
        raise HTTPException(status_code=500, detail="An error occurred while updating the fact.")

@router.post("/facts", response_model=FactResponse)
def create_fact(description: str, location_id: str, tags: list[str], current_user: User = Depends(get_current_user)):
    try:
        location = Location.objects.get(id=location_id)
    except DoesNotExist:
        logger.warning(f"Location ID: {location_id} does not exist")
        raise HTTPException(status_code=404, detail="Location not found")

    tag_objects = [Tag.objects.get(name=tag) for tag in tags]

    new_fact = Fact(
        description=description,
        location_id=location,
        user_id=current_user,
        tags=tag_objects
    )
    new_fact.save()
    logger.info(f"Fact created with ID: {new_fact.id} for location ID: {location_id}")
    return FactResponse(
        id=str(new_fact.id),
        description=new_fact.description,
        location_id=str(new_fact.location_id.id),
        user_id=str(current_user.id),
        created_at=new_fact.created_at
    )

File: routes/facts.js
const express = require("express");
const router = express.Router();
const { Fact, Location, Tag } = require("../models/models"); // Import Mongoose models
const { authenticateToken } = require("../utils/authenticate_token"); // Import the authenticateToken middleware
const logger = require("../utils/logger"); // Logger utility

// Get facts by location ID
router.get("/locations/:locationId/facts", authenticateToken, async (req, res) => {
  const { locationId } = req.params;
  const userId = req.user.id;

  logger.info(`Fetching facts for location ID: ${locationId} by user ID: ${userId}`);

  try {
    const facts = await Fact.find({ locationId }).populate("locationId").exec();
    if (!facts.length) {
      logger.warn(`No facts found for location ID: ${locationId}`);
      return res.status(404).json({ error: "No facts found for this location." });
    }

    logger.info(`Found ${facts.length} facts for location ID: ${locationId}`);
    const response = facts.map((fact) => ({
      id: fact._id,
      description: fact.description,
      locationId: fact.locationId._id,
      userId: fact.userId,
      createdAt: fact.createdAt,
    }));

    res.json(response);
  } catch (err) {
    logger.error(`Error fetching facts: ${err.message}`);
    res.status(500).json({ error: "Internal server error." });
  }
});

// Create a fact for a location
router.post("/locations/:locationId/facts", authenticateToken, async (req, res) => {
  const { locationId } = req.params;
  const { description, tags } = req.body;
  const userId = req.user.id;

  logger.info(`Creating fact for location ID: ${locationId} by user ID: ${userId}`);

  try {
    const location = await Location.findById(locationId);
    if (!location) {
      logger.warn(`Location ID: ${locationId} does not exist`);
      return res.status(404).json({ error: "Location not found." });
    }

    // Validate tags
    const tagObjects = [];
    for (const tag of tags) {
      const tagObject = await Tag.findOne({ name: tag });
      if (!tagObject) {
        logger.warn(`Tag '${tag}' not found.`);
        return res.status(400).json({ error: `Tag '${tag}' not found.` });
      }
      tagObjects.push(tagObject);
    }

    // Create and save the fact
    const newFact = new Fact({
      description,
      locationId,
      userId,
      tags: tagObjects,
    });
    await newFact.save();

    logger.info(`Fact created with ID: ${newFact._id} for location ID: ${locationId}`);
    res.status(201).json({
      id: newFact._id,
      description: newFact.description,
      locationId: newFact.locationId,
      userId: newFact.userId,
      createdAt: newFact.createdAt,
    });
  } catch (err) {
    logger.error(`Error creating fact: ${err.message}`);
    res.status(500).json({ error: "Internal server error." });
  }
});

// Delete a fact by ID
router.delete("/facts/:factId", authenticateToken, async (req, res) => {
  const { factId } = req.params;
  const userId = req.user.id;

  logger.info(`Attempting to delete fact with ID: ${factId} by user ID: ${userId}`);

  try {
    const fact = await Fact.findOneAndDelete({ _id: factId, userId });
    if (!fact) {
      logger.warn(`Fact ID: ${factId} does not exist or not owned by user ID: ${userId}`);
      return res.status(404).json({ error: "Fact not found." });
    }

    logger.info(`Fact with ID: ${factId} deleted`);
    res.status(204).send();
  } catch (err) {
    logger.error(`Error deleting fact: ${err.message}`);
    res.status(500).json({ error: "Internal server error." });
  }
});

// Update a fact by ID
router.put("/locations/:locationId/facts/:factId", authenticateToken, async (req, res) => {
  const { locationId, factId } = req.params;
  const { description } = req.body;
  const userId = req.user.id;

  logger.info(`Updating fact with ID: ${factId} for location ID: ${locationId} by user ID: ${userId}`);

  try {
    const location = await Location.findById(locationId);
    if (!location) {
      logger.warn(`Location ID ${locationId} does not exist.`);
      return res.status(404).json({ error: "Location not found." });
    }

    const fact = await Fact.findOneAndUpdate(
      { _id: factId, locationId, userId },
      { description },
      { new: true }
    );
    if (!fact) {
      logger.warn(`Fact with ID ${factId} not found for location ID ${locationId}`);
      return res.status(404).json({ error: "Fact not found." });
    }

    logger.info(`Fact with ID ${factId} updated successfully for location ID: ${locationId}`);
    res.json({
      id: fact._id,
      description: fact.description,
      locationId: fact.locationId,
      userId: fact.userId,
      createdAt: fact.createdAt,
    });
  } catch (err) {
    logger.error(`Error updating fact: ${err.message}`);
    res.status(500).json({ error: "Internal server error." });
  }
});

module.exports = router;
